### PR TITLE
feat: integrate nvim-web-devicons

### PR DESCRIPTION
**Pull request by spywhere**
This PR adds support for [nvim-web-devicons](https://github.com/kyazdani42/nvim-web-devicons). Mind that this only works with neovim with lua support (hence the check).

Let me know should I need to refactor anything, I'm very much new to the vimscript.

### CODE REVIEW

File: autoload/lightline/bufferline.vim [hunk 1]
```diff
@@ -58,6 +73,8 @@ function! s:get_buffer_name(i, buffer, path)
   endif
   if s:enable_devicons == 1 && exists('*WebDevIconsGetFileTypeSymbol')
     let l:name = WebDevIconsGetFileTypeSymbol(fnamemodify(bufname(a:buffer), ':t')) . ' ' . l:name
+  elseif s:enable_devicons == 1 && exists('g:loaded_devicons')
```
Comment (spywhere) [inline autoload/lightline/bufferline.vim]: Let me know if this flag should be put separately, but to me I think it's a good idea to use the same flag and implement a fallback instead. So this should take the vim-devicons first then nvim-web-devicons

**Comment by mengelbrecht** [comment]
I installed nvim-web-devicons and tried your changes with Neovim 0.4.4. However, no buffernames are displayed for me...
Am I missing a setup step?

**Comment by spywhere** [comment]
> I installed nvim-web-devicons and tried your changes with Neovim 0.4.4. However, no buffernames are displayed for me...
> Am I missing a setup step?

Ah, okay. Let me try testing more. Since I'm using the neovim on master build, that might be the case.

**Comment by spywhere** [comment]
Okay, so there seems to be only available on nvim 0.5 and later.

Screen Shot 2563-11-17 at 09 45 14

I'll make a fix to check for the version then.

Here's the complete configurations that I tested

```vim
if empty(glob('~/.config/nvim/autoload/plug.vim'))
  silent !curl -fLo ~/.config/nvim/autoload/plug.vim --create-dirs https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
  autocmd VimEnter * PlugInstall --sync | q | source $MYVIMRC
  let g:first_install = 1
else
  let g:init_vim_loaded = 1
endif

" Setup plugins
call plug#begin('~/.config/nvim/plugged')
Plug 'kyazdani42/nvim-web-devicons'
Plug 'itchyny/lightline.vim'
Plug 'mengelbrecht/lightline-bufferline'
call plug#end()

" Already loaded, stop processing
if !exists('g:init_vim_loaded')
  finish
endif

" Lightline-bufferline
let g:lightline#bufferline#enable_devicons = 1
let g:lightline#bufferline#min_buffer_count = 1

" Lightline
let g:lightline = {
\   'component_expand': {
\     'buffers': 'lightline#bufferline#buffers'
\   },
\ }

let g:lightline.tabline = {
\   'left': [
\     ['buffers']
\   ],
\   'right': [
\     []
\   ],
\ }
```

**Comment by spywhere** [comment]
Here's the fix one, let me know if I should change something...

Screen Shot 2563-11-17 at 09 47 00

**Comment by mengelbrecht** [comment]
Thanks @spywhere, it works for me too using neovim 0.5.

Is there a way to detect `nvim-web-devicons` specifically? 

The code relies on the fact that if `g:loaded_devicons = 1` is set but devicons is not installed then `nvim-web-devicons` must be installed. However, if `nvim-web-devicons` is not installed but another fork (hypothetically) is installed that also defines `g:loaded_devicons = 1` the bufferline will be gone since the code tries to use `nvim-web-devicons`.

**Comment by spywhere** [comment]
> Thanks @spywhere, it works for me too using neovim 0.5.
> 
> Is there a way to detect `nvim-web-devicons` specifically?
> 
> The code relies on the fact that if `g:loaded_devicons = 1` is set but devicons is not installed then `nvim-web-devicons` must be installed. However, if `nvim-web-devicons` is not installed but another fork (hypothetically) is installed that also defines `g:loaded_devicons = 1` the bufferline will be gone since the code tries to use `nvim-web-devicons`.

I can use `g:nvim_web_devicons` for that. However, if you look [here](https://github.com/kyazdani42/nvim-web-devicons/blob/master/plugin/nvim-web-devicons.vim#L7),I'm not sure if we should use that with the note there.

Anyway, let me update so it use `g:nvim_web_devicons` instead then.

**Comment by mengelbrecht** [comment]
Looks good, thanks!

**Comment by ahrex** [comment]
Hello!

I believe this breaks MacVim 8.2.1719 (166), the [current release](https://github.com/macvim-dev/macvim/releases).

I tested by copying the only function I found, `lightline#bufferline#_click_handler`, and renaming it `lightline#bufferline#test`.

You can see that when placed above the lua block, the function is found, as expected:
Screen Shot 2020-12-02 at 18 11 33

But when placed below the lua block, it's not found:
Screen Shot 2020-12-02 at 18 10 54

This stops MacVim from processing the rest of the file, breaking this plugin.

Could we figure out a way to resolve?

Thanks!

**Comment by mengelbrecht** [comment]
@ahrex thanks for the report. There is a bug in vim which causes this, see #77. It is fixed in vim 8.2.1910. However, I see that there is no MacVim release which includes that vim version yet.

I will revert the `nvim-web-devicons` feature for now since it breaks this plugin for all vim users with a version < 8.2.1910 (FYI: @spywhere)

**Comment by ahrex** [comment]
Thank you! Hopefully we can re-apply this change once enough Vim variants with the fix are released.

**Comment by mengelbrecht** [comment]
@ahrex @spywhere I reverted the `nvim-web-devicons` integration in 9dffe95e90e331faaf363e179970ffc46afaeb59.

**Comment by spywhere** [comment]
> @ahrex @spywhere I reverted the `nvim-web-devicons` integration in [9dffe95](https://github.com/mengelbrecht/lightline-bufferline/commit/9dffe95e90e331faaf363e179970ffc46afaeb59).

No worry! I'll kept my fork for my own use until then then.

**Pull request merged**